File: src/main/java/com/beat/global/external/notification/slack/event/MemberRegisteredEventListener.java
package com.beat.global.external.notification.slack.event;

import static com.beat.global.external.notification.slack.vo.SlackConstant.BRAND_COLOR;

import java.util.List;

import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;
import org.springframework.transaction.event.TransactionPhase;
import org.springframework.transaction.event.TransactionalEventListener;

import com.beat.domain.member.application.dto.event.MemberRegisteredEvent;
import com.beat.domain.member.port.in.MemberUseCase;
import com.beat.global.external.notification.slack.application.SlackService;
import com.beat.global.external.notification.slack.vo.block.DividerBlock;
import com.beat.global.external.notification.slack.vo.block.HeaderBlock;
import com.beat.global.external.notification.slack.vo.block.SectionBlock;
import com.beat.global.external.notification.slack.vo.message.SlackMessage;
import com.beat.global.external.notification.slack.vo.text.MarkdownText;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@Component
@RequiredArgsConstructor
public class MemberRegisteredEventListener {

	private final MemberUseCase memberUseCase;
	private final SlackService slackService;

	@Async("taskExecutor")
	@TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
	public void sendSlackNotification(MemberRegisteredEvent event) {
		try {
			slackService.sendToMemberChannel(buildMessage(event));
		} catch (Exception e) {
			log.error("Slack 전송 실패 - 닉네임: {}", event.nickname(), e);
		}
	}

	private SlackMessage buildMessage(MemberRegisteredEvent event) {
		long memberCount = memberUseCase.countMembers();
		return SlackMessage.newInstance(
			List.of(
				HeaderBlock.newInstance("🎉 BEAT 신규 회원 가입 🎉"),
				SectionBlock.newInstanceWithFields(List.of(
					MarkdownText.newInstance("*🙋 닉네임*\n" + event.nickname()),
					MarkdownText.newInstance("*👥 누적 회원 수*\n" + memberCount + "명")
				)),
				DividerBlock.newInstance()
			),
			BRAND_COLOR
		);
	}
}

File: application/frontoffice/src/main/kotlin/com/beat/application/frontoffice/member/command/MemberRegisteredEventListener.kt
package com.beat.application.frontoffice.member.command

import com.beat.domain.member.repository.MemberRepository
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Component
import org.springframework.transaction.event.TransactionPhase
import org.springframework.transaction.event.TransactionalEventListener

@Component
internal class MemberRegisteredEventListener(
    private val memberRepository: MemberRepository,
    private val memberRegistrationNotifier: MemberRegistrationNotifier,
) {
    @Async("beatAsyncExecutor")
    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
    fun sendSlackNotification(event: MemberRegisteredEvent) {
        try {
            memberRegistrationNotifier.send(
                MemberRegistrationNotification(
                    nickname = event.nickname,
                    memberCount = memberRepository.count(),
                )
            )
        } catch (exception: RuntimeException) {
            log.error(exception) {
                "Member registration Slack notification failed: errorType=${exception.javaClass.simpleName}"
            }
        }
    }

    private companion object {
        val log = KotlinLogging.logger {}
    }
}
